### PR TITLE
Custom expression: improve the type-checker

### PR DESCRIPTION
**Pull request by ariya**
Use an improved semantic validation layered on top of the existing resolver (around ~40 LoC) instead of its own dedicated logic (~240 LoC).  Not only the improved type-checker is leaner and faster, it's also more accurate in some corner cases.

The type-checking process is also deferred to the compilation stage, instead of earlier in the parsing stage.

An example where this type checker spots a problem is in the repro for issue #15714. The repro test incorrectly attempted to check the usage of `percentile` in the custom column. This new type checker wasn't happy about that since the said function can be used only for aggregation. The repro is tweaked accordingly as part of this PR (and moved to `scenarios/questions` as well).

Additionally some other tests need to be adjusted because otherwise they triggered the same type of complaints.

Meanwhile, everything else, should work as expected.

### CODE REVIEW

File: frontend/test/metabase/scenarios/question/reproductions/15714-cc-postgres-percentile-accepts-two-params.cy.spec.js [hunk 1]
```diff
@@ -14,11 +14,12 @@ describe("postgres > question > custom columns", () => {
   });
 
   it("`Percentile` custom expression function should accept two parameters (metabase#15714)", () => {
-    cy.icon("add_data").click();
+    cy.findByText("Pick the metric you want to see").click();
```
Comment (flamber) [inline frontend/test/metabase/scenarios/question/reproductions/15714-cc-postgres-percentile-accepts-two-params.cy.spec.js:17]: Didn't know that class was Custom Column. At least now it's more clear.

**Review by flamber — APPROVED** [review]
There's a lot of other stuff going on and I don't quite understand all of it.

File: frontend/src/metabase/lib/expressions/resolver.js [hunk 2]
```diff
@@ -35,7 +37,12 @@ function findMBQL(op) {
   return clause;
 }
 
-export function resolve(expression, type, fn) {
+const isCompatible = (a, b) =>
+  a === b ||
```
Comment (akiselev) [inline frontend/src/metabase/lib/expressions/resolver.js:41]: nit(?): These are three semantically different cases so it feels iffy putting them in a single expression. If statements without braces are also "not great" but I think in this case might make it clearer that there are three different cases to look for.

**Review by akiselev — APPROVED** [review]

File: frontend/src/metabase/lib/expressions/resolver.js (hunk 2 above)
Comment (ariya) [reply frontend/src/metabase/lib/expressions/resolver.js]: Good point, thanks! I've changed it so the code becomes a bit more verbose, but I agree that clarity is important here.

**Pull request merged**